### ISSUE TITLE
Opaque crash in `Number::set_overwrite()`

### ISSUE THREAD

**Issue by novafacing** [trailofbits/maat#49]
I'm getting a crash trying to simply run a binary with an input concolic buffer following the tutorial [here](http://maat.re/tutorials/dse.html). I'm running against the binary `AIS-Lite` from the CGC dataset, and I've attached a .tar.gz of the directory with the binary and libraries. A minimized version of my script is below to reproduce the issue.

Any help would be appreciated to figure out what I'm doing wrong here, it would also be super helpful to have a more detailed error explanation or traceback on errors like this!

Thanks for the hard work folks, this system looks really promising once it matures a bit!

[AIS-Lite.tar.gz](https://github.com/trailofbits/maat/files/8129287/AIS-Lite.tar.gz)

```python3
from pathlib import Path
from argparse import ArgumentParser
from typing import List, Optional
from maat import MaatEngine, OS, ARCH, BIN
from angr import Project
from logging import basicConfig, getLogger

basicConfig()
logger = getLogger(__name__)


class TestMaat:
    """Can ma'at replace Triton?"""

    BASE = 0x400000

    def __init__(
        self,
        binary: Path,
        input_file: Path,
        args: Optional[List[str]] = None,
        libdirs: Optional[List[str]] = None,
    ) -> None:
        """
        Set up the maat engine with our binary.

        :param binary: Path to the binary to run
        :param args: Arguments to pass to the binary
        """
        self.binary = binary
        self.input_file = input_file
        self.args = args
        self.libdirs = (
            list(map(lambda l: str(l.resolve()), libdirs))
            if libdirs is not None
            else []
        )
        self.engine = MaatEngine(ARCH.X64, OS.LINUX)
        assert binary.is_file(), f"{binary} is not a file or doesn't exist."

        logger.info(f"Loading binary {self.binary} with libdirs {self.libdirs}")

        self.engine.load(
            str(binary.resolve()),
            BIN.ELF64,
            base=self.BASE,
            args=args if args is not None else [],
            libdirs=self.libdirs,
        )

        self.set_input()

    def run(self) -> None:
        """
        Run the binary.
        """
        self.engine.run()

    def set_input(self) -> None:
        """
        Set up the input for the binary.
        """

        stdin = self.engine.env.fs.get_fa_by_handle(0)  # Get stdin simfile
        contents = self.input_file.read_bytes()
        input_buffer = self.engine.vars.new_concolic_buffer(
            "stdin",
            contents,
            len(contents),
        )
        stdin.write_buffer(input_buffer)


if __name__ == "__main__":
    parser = ArgumentParser(prog="repro")
    parser.add_argument(
        "--binary", type=Path, required=True, help="Path to binary to execute."
    )
    parser.add_argument(
        "--input", type=Path, required=True, help="Path to input file to read."
    )
    parser.add_argument("--args", nargs="*", help="Arguments to pass to binary.")
    parser.add_argument(
        "--libdirs", nargs="*", type=Path, help="Library directories to load."
    )
    cli_args = parser.parse_args()

    stage1 = TestMaat(
        cli_args.binary,
        cli_args.input,
        cli_args.args,
        cli_args.libdirs + [Path("/lib/x86_64-linux-gnu/")],
    )
    stage1.run()
```

**Comment by novafacing** [comment]
Actually, I thought it was setting the input that causes the crash, but if `self.set_input()` is commented out, it still throws the exception:
```
terminate called after throwing an instance of 'maat::expression_exception'
  what():  Number::set_overwrite(): src number is too big to fit in dest!
Aborted (core dumped)
```

**Comment by Boyan-MILANOV** [comment]
Hey, thanks for reporting this bug. I'll look into it!

**Comment by Boyan-MILANOV** [comment]
I've fixed the bug in #50. The issue was that the `PMINUB` instruction wasn't correctly emulated.

One mistake I spotted in your example: `AIS-Lite` is not position-independent so you shouldn't specify the load address with `base=self.BASE` when calling `engine.load()`. As stated in the [doc](https://maat.re/python_api/class_MaatEngine.html#85869075) for `load()`: `base | (Optional) Base address where to relocate the binary if position-independent`.  Calling with `base=self.BASE` adds the base to the already-present base and results in the program being loaded at `0x800000` instead of `0x400000`. We should probably throw an error when users use `base` with a non-position-independant binary. I'll open a dedicated issue.

The issue is fixed but the binary still doesn't run properly because `libtiny-AES128-C.so` isn't loaded properly in the emulation environment. That's because `libtiny-AES128-C.so` is a dependency of `libgcg.so` while Maat loads only the direct dependencies of `AIS-Lite`. I'll see how hard it would be to recursively scan and add shared libraries accross shared libraries.

**Comment by Boyan-MILANOV** [comment]
@novafacing: this is working for me now. On my machine the execution stops at `AIS-LITE: symbol lookup error: AIS-LITE: undefined symbol: __libc_start_main, version GLIBC_2.2.5` because I haven't the correct `libc` version for your compiled binary. But it should work for you locally. 

We'll close this issue by merging #50, if you get more errors running the challenge just open new issues.
